### PR TITLE
#14 - crud for news

### PR DESCRIPTION
**Pull request by MrHDOLEK**

### CODE REVIEW

File: resources/views/news/index.blade.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        <div>
+            <h1>News</h1>
+            @auth
+                <a href="{{ route('news.create') }}">Create new news</a>
+            @endauth
+            @if ($news->count())
+                @foreach ($news as $singleNews)
+                    <div>
+                        {{ $singleNews->title }}
+                        {!! \Illuminate\Mail\Markdown::parse($singleNews->content) !!}
```
Comment (kamilpiech97) [inline resources/views/news/index.blade.php:14]:
```suggestion
                        {!! Markdown::parse($singleNews->content) !!}
```
make alias for that.

File: src/Http/Controllers/NewsController.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    public function edit(News $news): View
+    {
+        return view("news.edit")
+            ->with("news", $news);
+    }
+
+    public function update(StoreNewsRequest $request, News $news): RedirectResponse
+    {
+        $news->update($request->validated());
```
Comment (kamilpiech97) [inline src/Http/Controllers/NewsController.php:48]: On `edit` u did not clean content like in `store` method?

File: src/Http/Controllers/NewsController.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return view("news.create");
+    }
+
+    public function store(StoreNewsRequest $request): RedirectResponse
+    {
+        $input = $request->validated();
+        $input["content"] = Purifier::clean($input["content"]);
+        $input["slug"] = Str::slug($input["title"]);
+
```
Comment (kamilpiech97) [inline src/Http/Controllers/NewsController.php:34]: Maybe prepare data in request class?

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/NewsController.php (hunk 2 above)
Comment (agawle) [inline src/Http/Controllers/NewsController.php:48]: Same as news deletion, anyone can update any news

**Review by agawle — CHANGES_REQUESTED** [review]

File: src/Http/Routing/WebRouting.php [hunk 4]
```diff
@@ -20,11 +20,13 @@ public function wire(): void
...
-        $this->router->get("/auth/login", [LoginController::class, "store"])->name("login.form");
-        $this->router->post("/auth/login", [LoginController::class, "login"])->name("login");
-        $this->router->get("/auth/logout", [LoginController::class, "logout"])->name("logout")->middleware("auth");
+        $this->router->controller(RegisterController::class)->group(function (): void {
+            $this->router->get("/auth/register", "create")->name("register.form");
+            $this->router->post("/auth/register", "store")->name("register");
+            $this->router->get("/auth/login", "store")->name("login.form");
+            $this->router->post("/auth/login", "login")->name("login");
+            $this->router->get("/auth/logout", "logout")->name("logout")->middleware("auth");
+        });
```
Comment (Majkulnn) [inline src/Http/Routing/WebRouting.php:29]: Why you change that to one controller group?

File: resources/views/news/index.blade.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                        {{ $singleNews->author }}
+
+                        <a href="{{ route('news.edit', $singleNews) }}">Edit</a>
+                        <form action="{{ route('news.destroy', $singleNews) }}" method="post">
+                            @csrf
+                            @method('DELETE')
+                            <button type="submit"  onclick="return confirm('Delete News? This operation is irreversible.')">Delete</button>
```
Comment (mtracz) [inline resources/views/news/index.blade.php:21]: So do you send `post` or `delete` form?

**Review by mtracz — CHANGES_REQUESTED** [review]

File: resources/views/news/index.blade.php (hunk 5 above)
Comment (mtracz) [reply resources/views/news/index.blade.php:21]: Ok, got it. Regarding to https://laravel.com/docs/9.x/blade#method-field, there is no question.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/NewsController.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public function update(StoreNewsRequest $request, News $news): RedirectResponse
+    {
+        $news->update($request->validated());
+
+        return redirect()->route("news");
+    }
+
+    public function destroy(News $news): RedirectResponse
```
[the remark this answers is not in the document]
Comment (agawle) [reply src/Http/Controllers/NewsController.php:49]: I've created an issue  for that then. #79

File: database/migrations/2022_03_18_195003_create_news_table.php [hunk 7]
```diff
@@ -11,9 +11,9 @@ public function up(): void
     {
         Schema::create("news", function (Blueprint $table): void {
             $table->id();
-            $table->foreignId("author")->constrained()->onDelete("cascade");
+            $table->foreignId("author_id")->constrained()->onDelete("cascade");
             $table->string("title");
-            $table->longText("content");
+            $table->longText("text");
```
[the remark this answers is not in the document]
Comment (MrHDOLEK) [reply database/migrations/2022_03_18_195003_create_news_table.php:16]: Why ? Like it's not even in production or in daily use by outside users.

[the remark this answers is not in the document]
Comment (krzysztofrewak) [reply database/migrations/2022_03_18_195003_create_news_table.php]: There's a lot of people just reviewing development. I don't want to check if I need to rebuild the whole database every time I'm pulling any changes from Github.

File: resources/views/meetups/index.blade.php [hunk 8]
```diff
@@ -17,7 +17,7 @@
                         {{ $meetup->language }}
 
                         <a href="{{ route('meetups.edit', $meetup) }}">Edit</a>
-                        <form action="{{ route('meetups.destroy', $meetup) }}" method="post">
+                        <form action="{{ route('meetups.destroy', $meetup) }}">
```
Comment (krzysztofrewak) [inline resources/views/meetups/index.blade.php:20]: This won't work.

File: config/purifier.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
+<?php
```
Comment (krzysztofrewak) [inline config/purifier.php:1]: Did you talk with anyone about adding this dependency here?

File: resources/views/meetups/index.blade.php [hunk 10]
```diff
@@ -17,7 +17,7 @@
                         {{ $meetup->language }}
```
Comment (krzysztofrewak) [inline resources/views/meetups/index.blade.php:17]: Why did you changed chmods from 644 to 755 for so many files?

File: src/Models/News.php [hunk 11]
```diff
@@ -22,8 +23,11 @@ class News extends Model
     use HasFactory;
 
     protected $fillable = [
+        "author_id",
+        "slug",
+        "title",
         "name",
-        "content",
+        "text",
```
Comment (krzysztofrewak) [inline src/Models/News.php:30]: Why did you change this name? And why did you not change docblock?

File: resources/js/app.js [hunk 12]
```diff
@@ -1,3 +1,30 @@
...
+import '@toast-ui/editor/dist/toastui-editor.css';
+
+const editor = new Editor({
+    el: document.querySelector('#editor'),
+    height: '400px',
+    initialEditType: 'markdown',
+    placeholder: 'Write something cool!',
+})
+
+if (document.querySelector('#createNews')) {
```
Comment (krzysztofrewak) [inline resources/js/app.js:12]: Styling of your Javascript files is a mess. Please look into our eslint in other projects like here https://github.com/blumilksoftware/toby/blob/main/.eslintrc.js

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: resources/views/meetups/index.blade.php (hunk 8 above)
Comment (MrHDOLEK) [reply resources/views/meetups/index.blade.php:20]: I tested and work

File: config/purifier.php (hunk 9 above)
Comment (MrHDOLEK) [reply config/purifier.php:1]: not because I thought it was obvious

File: resources/js/app.js (hunk 12 above)
Comment (MrHDOLEK) [reply resources/js/app.js:12]: I know of a eslint but for this project is not setup a eslint . A separate pull req for me to configure the eslint

File: src/Models/News.php (hunk 11 above)
Comment (MrHDOLEK) [reply src/Models/News.php:30]: My mistake

File: resources/views/meetups/index.blade.php (hunk 10 above)
Comment (MrHDOLEK) [reply resources/views/meetups/index.blade.php:17]: I did not change the chmod and it seems to me that the last pull req has been changed on the master

File: config/purifier.php (hunk 9 above)
Comment (krzysztofrewak) [reply config/purifier.php:1]: What was obvious?

File: database/migrations/2022_04_26_204417_rename_author_to_author_id_news_table.php [hunk 13]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class() extends Migration {
```
Comment (krzysztofrewak) [inline database/migrations/2022_04_26_204417_rename_author_to_author_id_news_table.php:9]: You're creating a fix migration to migration that wasn't even committed to main? Oh, come on.

File: resources/js/app.js (hunk 12 above)
Comment (krzysztofrewak) [reply resources/js/app.js:12]: Then fix styling here manually, please.

File: resources/views/welcome.blade.php [hunk 14]
```diff
@@ -0,0 +1,132 @@
+<!DOCTYPE html>
```
Comment (krzysztofrewak) [inline resources/views/welcome.blade.php:1]: Why did you restore this default view?

File: src/Http/Requests/News/NewsRequest.php [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        return [
+            "title" => "required|string",
+            "text" => "required|string",
+        ];
+    }
+
+    protected function prepareForValidation(): void
+    {
+        $cleanText = Purifier::clean($this->text);
```
Comment (krzysztofrewak) [inline src/Http/Requests/News/NewsRequest.php:22]: Please explain here what this purifier do.

File: src/Models/News.php [hunk 16]
```diff
@@ -33,4 +37,12 @@ public function user(): BelongsTo
     {
         return $this->belongsTo(User::class);
     }
+
+    protected static function boot(): void
+    {
+        parent::boot();
+        static::creating(function ($news): void {
```
Comment (krzysztofrewak) [inline src/Models/News.php:44]: Typing.

**Review by krzysztofrewak — CHANGES_REQUESTED** [review]

File: database/migrations/2022_04_26_204417_rename_author_to_author_id_news_table.php (hunk 13 above)
Comment (agawle) [reply database/migrations/2022_04_26_204417_rename_author_to_author_id_news_table.php:9]: @krzysztofrewak I think he created that because per @kamilpiech97 request https://github.com/blumilksoftware/meetup/pull/39#discussion_r858268182

File: src/Http/Requests/News/NewsRequest.php (hunk 15 above)
Comment (MrHDOLEK) [reply src/Http/Requests/News/NewsRequest.php:22]: Purifier cleans from html scripts and leaves only md but create html <p> tag and I haven't seen in documentation to change it

File: config/purifier.php (hunk 9 above)
Comment (MrHDOLEK) [reply config/purifier.php:1]: adding config to a package that requires it

File: src/Http/Requests/News/NewsRequest.php (hunk 15 above)
Comment (MrHDOLEK) [reply src/Http/Requests/News/NewsRequest.php:22]: md

Comment (MrHDOLEK) [reply src/Http/Requests/News/NewsRequest.php:22]: i mean clear md can't include js and html

File: webpack.mix.js [hunk 17]
```diff
@@ -1,6 +1,7 @@
 let mix = require('laravel-mix');
 
-mix
+mix.browserSync('127.0.0.1');
+mix.disableSuccessNotifications()
```
Comment (agawle) [inline webpack.mix.js:4]: Why those lines are duplicated with 11-12?

File: resources/views/news/index.blade.php [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+                @foreach ($news as $singleNews)
+                    <div>
+                        {{ $singleNews->title }}
+                        {!! Str::markdown($singleNews->text) !!}
+                        {{ $singleNews->author }}
+
+                        <a href="{{ route('news.edit', $singleNews) }}">Edit</a>
+                        <form action="{{ route('news.destroy', $singleNews) }}" method="post">
+                            @csrf
+                            @method('DELETE')
```
Comment (agawle) [inline resources/views/news/index.blade.php:20]: If you removed method="post" from meetups/index view you should either delete it here, if that's working, or write it back in meetups.

**Review by agawle — CHANGES_REQUESTED** [review]

File: database/migrations/2022_03_18_195003_create_news_table.php [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        Schema::create("news", function (Blueprint $table): void {
+            $table->id();
+            $table->foreignId("author")->constrained()->onDelete("cascade");
```
[the remark this answers is not in the document]
Comment (MrHDOLEK) [reply database/migrations/2022_03_18_195003_create_news_table.php]: I rename author to author_id

File: resources/views/news/index.blade.php (hunk 18 above)
Comment (agawle) [reply resources/views/news/index.blade.php]: I still think that keeping the method="post" is a better way, but as long as you will be consistent in all forms and if that really works, i guess it's ok @MrHDOLEK .

File: webpack.mix.js (hunk 17 above)
Comment (agawle) [reply webpack.mix.js:4]: @MrHDOLEK i still don't know why it's duplicated.

File: src/Http/Routing/WebRouting.php [hunk 20]
```diff
@@ -89,6 +95,15 @@ public function wire(): void
             $this->router->delete("/speakers/{speaker}", "destroy")->name("speakers.destroy");
         });
 
+        $this->router->controller(NewsController::class)->group(function (): void {
+            $this->router->get("/news", "index")->name("news");
+            $this->router->get("/news/create", "create")->name("news.create");
```
Comment (agawle) [inline src/Http/Routing/WebRouting.php:100]: Please attach middleware auth where needed, or at least to news creation, as you are trying to attach a news to a user, so it would be nice to make sure, that the user is signed in

**Review by agawle — CHANGES_REQUESTED** [review]

File: resources/views/news/index.blade.php [hunk 21]
```diff
@@ -15,7 +15,7 @@
                         {{ $singleNews->author }}
 
                         <a href="{{ route('news.edit', $singleNews) }}">Edit</a>
-                        <form action="{{ route('news.destroy', $singleNews) }}" method="post">
+                        <form action="{{ route('news.destroy', $singleNews) }}">
```
Comment (agawle) [inline resources/views/news/index.blade.php:18]: I would still prefer keeping the method="post" in cases where request isn't a get request, but if that's working, then just remember to keep it consistent across all forms.

Comment (MrHDOLEK) [reply resources/views/news/index.blade.php:18]: Maybe create issue for delete method="post" ?

**Comment by krzysztofrewak** [comment]
Please resolve conflicts before further code review.

File: src/Http/Routing/WebRouting.php [hunk 22]
```diff
@@ -93,6 +94,15 @@ public function wire(): void
             $this->router->delete("/speakers/{speaker}", "destroy")->name("speakers.destroy");
         });
 
+        $this->router->controller(NewsController::class)->group(function (): void {
+            $this->router->get("/news", "index")->name("news");
+            $this->router->get("/news/create", "create")->middleware("auth")->name("news.create");
+            $this->router->post("/news", "store")->name("news.store");
```
Comment (agawle) [inline src/Http/Routing/WebRouting.php:100]: Please also attach middleware here, as in the store method you are trying to get a user from request

**Review by agawle — CHANGES_REQUESTED** [review]

File: src/Http/Routing/WebRouting.php (hunk 22 above)
Comment (MrHDOLEK) [reply src/Http/Routing/WebRouting.php]: done

File: resources/views/meetups/index.blade.php (hunk 10 above)
Comment (MrHDOLEK) [reply resources/views/meetups/index.blade.php:17]: done i revert to chmod 644

**Review by agawle — APPROVED** [review]
